Mathlib/NumberTheory/AbelSummation.lean

```diff
@@ -0,0 +1,173 @@
+  exact (IntegrableOn.mono_set (hf_int.mul_const (∑ k ∈ Icc 0 n, c k))
+    (Set.Icc_subset_Icc h₃ h₄)).congr
+      <| ae_restrict_of_ae_restrict_of_subset
+        (Set.Icc_subset_Icc h₁ h₂) <| (ae_restrict_iff' measurableSet_Icc).mpr
+          (by filter_upwards [sumlocc c n] with t h ht using by rw [h ht])
+
+/-- Abel's summation formula. -/
+theorem _root_.sum_mul_eq_sub_sub_integral_mul (ha : 0 ≤ a) (hab : a ≤ b)
+    (hf_diff : ∀ t ∈ Set.Icc a b, DifferentiableAt ℝ f t)
+    (hf_int : IntegrableOn (deriv f) (Set.Icc a b)) :
+    ∑ k ∈ Ioc ⌊a⌋₊ ⌊b⌋₊, f k * c k =
+      f b * (∑ k ∈ Icc 0 ⌊b⌋₊, c k) - f a * (∑ k ∈ Icc 0 ⌊a⌋₊, c k) -
+        ∫ t in Set.Ioc a b, deriv f t * (∑ k ∈ Icc 0 ⌊t⌋₊, c k) := by
+  rw [← integral_of_le hab]
+  -- We prove some inequalities to be used later on by linarith
+  have : ⌊a⌋₊ ≤ a := Nat.floor_le ha
+  have : a < ⌊a⌋₊ + 1 := Nat.lt_floor_add_one _
+  have : b < ⌊b⌋₊ + 1 := Nat.lt_floor_add_one _
+  -- We consider two cases depending on whether the sum is empty or not
+  obtain hb | hb := eq_or_lt_of_le (Nat.floor_le_floor hab)
+  · rw [hb, Ioc_eq_empty_of_le le_rfl, sum_empty, ← sub_mul, integralmulsum c hf_diff hf_int,
+      sub_self]
+    any_goals linarith
+    rwa [← hb]
+  -- Some more inequalities for linarith
+  · have : 1 ≤ b := Nat.floor_pos.mp (by linarith)
+    have : ⌊b⌋₊ ≤ b := Nat.floor_le (by positivity)
+    have : ⌊a⌋₊ + 1 ≤ b := by rwa [← Nat.cast_add_one,  ← Nat.le_floor_iff (by positivity)]
+    have : a < ⌊b⌋₊ := by rwa [← Nat.floor_lt ha]
+    simp_rw [← smul_eq_mul, sum_Ioc_by_parts (fun k ↦ f k) _ hb, range_eq_Ico, Nat.Ico_succ_right,
+      smul_eq_mul]
+    rw [show ∑ k ∈ Ioc ⌊a⌋₊ (⌊b⌋₊ - 1), (f ↑(k + 1) - f ↑k) * ∑ n ∈ Icc 0 k, c n =
+          ∑ k ∈ Ico (⌊a⌋₊ + 1) ⌊b⌋₊, ∫ (t : ℝ) in ↑k..↑(k + 1), deriv f t * ∑ n ∈ Icc 0 ⌊t⌋₊, c n by
+      rw [← Nat.Ico_succ_succ, Nat.succ_eq_add_one,  Nat.succ_eq_add_one, Nat.sub_add_cancel
+        (by linarith)]
+      refine sum_congr rfl fun k hk ↦ (integralmulsum c hf_diff hf_int _ _ _ ?_ ?_ ?_ ?_ ?_).symm
+      all_goals simp [ineqofmemIco' hk]]
+    rw [sum_integral_adjacent_intervals_Ico (by linarith),
+      Nat.cast_add, Nat.cast_one, ← integral_interval_sub_left (a := a) (c := ⌊a⌋₊ + 1),
+      ← integral_add_adjacent_intervals (b := ⌊b⌋₊) (c := b), integralmulsum c hf_diff hf_int a
+      (⌊a⌋₊ + 1) ⌊a⌋₊, integralmulsum c hf_diff hf_int ⌊b⌋₊ b ⌊b⌋₊]
+    · ring
+    -- Now, we just need to check all the technical conditions
+    any_goals linarith
+    any_goals
+      (rw [intervalIntegrable_iff_integrableOn_Icc_of_le (by linarith)];
+        exact (integrablemulsum c ha hb hf_int).mono_set <|
+          (Set.Icc_subset_Icc_iff (by linarith)).mpr ⟨by linarith, by linarith⟩)
+    · refine fun k hk ↦ (intervalIntegrable_iff_integrableOn_Icc_of_le (by simp)).mpr
+        <| (integrablemulsum c ha hb hf_int).mono_set <|
+          (Set.Icc_subset_Icc_iff (by simp)).mpr ⟨?_, ?_⟩
+      all_goals simp [ineqofmemIco hk]
```

Comment:
Here's a version which compiles 3.5x quicker, at the cost of slightly less automation:
```suggestion
/-- Abel's summation formula. -/
theorem _root_.sum_mul_eq_sub_sub_integral_mul (ha : 0 ≤ a) (hab : a ≤ b)
    (hf_diff : ∀ t ∈ Set.Icc a b, DifferentiableAt ℝ f t)
    (hf_int : IntegrableOn (deriv f) (Set.Icc a b)) :
    ∑ k ∈ Ioc ⌊a⌋₊ ⌊b⌋₊, f k * c k =
      f b * (∑ k ∈ Icc 0 ⌊b⌋₊, c k) - f a * (∑ k ∈ Icc 0 ⌊a⌋₊, c k) -
        ∫ t in a..b, deriv f t * (∑ k ∈ Icc 0 ⌊t⌋₊, c k) := by
  -- We prove some inequalities to be used later on by linarith
  have aux1 : ⌊a⌋₊ ≤ a := Nat.floor_le ha
  have aux2 : b ≤ ⌊b⌋₊ + 1 := (Nat.lt_floor_add_one _).le
  -- We consider two cases depending on whether the sum is empty or not
  obtain hb | hb := eq_or_lt_of_le (Nat.floor_le_floor hab)
  · rw [hb, Ioc_eq_empty_of_le le_rfl, sum_empty, ← sub_mul,
      integralmulsum c hf_diff hf_int _ _ ⌊b⌋₊ hab (hb ▸ aux1) aux2 le_rfl le_rfl, sub_self]
  -- Some more inequalities for linarith
  have aux3 : a ≤ ⌊a⌋₊ + 1 := (Nat.lt_floor_add_one _).le
  have aux4 : ⌊a⌋₊ + 1 ≤ b := by rwa [← Nat.cast_add_one,  ← Nat.le_floor_iff (ha.trans hab)]
  have aux5 : ⌊b⌋₊ ≤ b := Nat.floor_le (ha.trans hab)
  have aux6 : a ≤ ⌊b⌋₊ := Nat.floor_lt ha |>.mp hb |>.le
  simp_rw [← smul_eq_mul, sum_Ioc_by_parts (fun k ↦ f k) _ hb, range_eq_Ico, Nat.Ico_succ_right,
    smul_eq_mul]
  have : ∑ k ∈ Ioc ⌊a⌋₊ (⌊b⌋₊ - 1), (f ↑(k + 1) - f k) * ∑ n ∈ Icc 0 k, c n =
        ∑ k ∈ Ico (⌊a⌋₊ + 1) ⌊b⌋₊, ∫ t in k..↑(k + 1), deriv f t * ∑ n ∈ Icc 0 ⌊t⌋₊, c n := by
    rw [← Nat.Ico_succ_succ, Nat.succ_eq_add_one,  Nat.succ_eq_add_one, Nat.sub_add_cancel
      (by omega), Eq.comm]
    exact sum_congr rfl fun k hk ↦ (integralmulsum c hf_diff hf_int _ _ _  (mod_cast k.le_succ)
      le_rfl (mod_cast le_rfl) (ineqofmemIco' hk).1 <| mod_cast (ineqofmemIco' hk).2)
  rw [this, sum_integral_adjacent_intervals_Ico hb, Nat.cast_add, Nat.cast_one,
    ← integral_interval_sub_left (a := a) (c := ⌊a⌋₊ + 1),
    ← integral_add_adjacent_intervals (b := ⌊b⌋₊) (c := b),
    integralmulsum c hf_diff hf_int _ _ _ aux3 aux1 le_rfl le_rfl aux4,
    integralmulsum c hf_diff hf_int _ _ _ aux5 le_rfl aux2 aux6 le_rfl]
  · ring
  -- now deal with the integrability side goals
  -- (Note we have 5 goals, but the 1st and 3rd are identical. TODO: find a non-hacky way of dealing
  -- with both at once.)
  · rw [intervalIntegrable_iff_integrableOn_Icc_of_le aux6]
    exact (integrablemulsum c ha hb hf_int).mono_set (Set.Icc_subset_Icc_right aux5)
  · rw [intervalIntegrable_iff_integrableOn_Icc_of_le aux5]
    exact (integrablemulsum c ha hb hf_int).mono_set (Set.Icc_subset_Icc_left aux6)
  · rw [intervalIntegrable_iff_integrableOn_Icc_of_le aux6]
    exact (integrablemulsum c ha hb hf_int).mono_set (Set.Icc_subset_Icc_right aux5)
  · rw [intervalIntegrable_iff_integrableOn_Icc_of_le aux3]
    exact (integrablemulsum c ha hb hf_int).mono_set (Set.Icc_subset_Icc_right aux4)
  · exact fun k hk ↦ (intervalIntegrable_iff_integrableOn_Icc_of_le (mod_cast k.le_succ)).mpr
      <| (integrablemulsum c ha hb hf_int).mono_set
        <| (Set.Icc_subset_Icc_iff (mod_cast k.le_succ)).mpr <| mod_cast (ineqofmemIco hk)
```
Most of the speedup comes from 2 things: firstly, `all_goals`, `any_goal` etc can take a while to fail on goals that don't fit, particularly if the goal (and tactic block) are complicated; secondly, `linarith` isn't all that good at recognising _fast_ if the goal is already in the context, which is actually true for every single one of the `linarith` calls in this proof as written.

Reply:
When I started this proof, I found out there was many technical side conditions that needed to be checked and it was getting a bit cumbersome, thus I decided to see if it was possible to just postponed them all and try to close them automatically at the end with some tactic. I guess I went a bit overboard and ended up solving trivial facts using `linearity`. I did notice in the end that the proof was getting a bit slow. 

Anyway, thanks for the improvements!